code/modules/modular_computers/file_system/programs/ntmessenger.dm

```diff
@@ -0,0 +1,352 @@
+				return
+
+			var/list/targets = list()
+
+			for(var/obj/item/modular_computer/mc in GetViewableDevices())
+				targets += mc
+
+			if(targets.len > 0)
+				send_message(usr, targets, TRUE)
+
+			return(UI_UPDATE)
+		if("PDA_sendMessage")
+			if(!sAndR)
+				to_chat(usr, span_notice("ERROR: Device has sending disabled."))
+				return
+			var/obj/item/modular_computer/target = locate(params["ref"])
+			if(!target)
+				return // we don't want tommy sending his messages to nullspace
+			if(!(target.saved_identification == params["name"] && target.saved_job == params["job"]))
+				to_chat(usr, span_notice("ERROR: User no longer exists."))
+				return
+
+			var/obj/item/computer_hardware/hard_drive/drive = target.all_components[MC_HDD]
+
+			for(var/datum/computer_file/program/messenger/app in drive.stored_files)
+				if(!app.sAndR && !sending_virus)
+					to_chat(usr, span_notice("ERROR: Device has receiving disabled."))
+					return
+				if(sending_virus)
+					var/obj/item/computer_hardware/hard_drive/role/virus/disk = computer.all_components[MC_HDD_JOB]
+					if(istype(disk))
+						disk.send_virus(target, usr)
+						return(UI_UPDATE)
+				send_message(usr, list(target))
+				return(UI_UPDATE)
+		if("PDA_clearPhoto")
+			computer.saved_image = null
+			photo_path = null
+			return(UI_UPDATE)
+		if("PDA_toggleVirus")
+			sending_virus = !sending_virus
+			return(UI_UPDATE)
+
+
+/datum/computer_file/program/messenger/ui_data(mob/user)
+	var/list/data = get_header_data()
+
+	var/obj/item/computer_hardware/hard_drive/role/disk = computer.all_components[MC_HDD_JOB]
+
+	data["owner"] = computer.saved_identification
+	data["messages"] = messages
+	data["ringerStatus"] = ringerStatus
+	data["sAndR"] = sAndR
+	data["messengers"] = ScrubMessengerList()
+	data["viewingMessages"] = viewingMessages
+	data["sortByJob"] = sort
+	data["isSilicon"] = is_silicon
+	data["photo"] = photo_path
+
+	if(disk)
+		data["canSpam"] = disk.CanSpam()
+		data["virus_attach"] = istype(disk, /obj/item/computer_hardware/hard_drive/role/virus)
+		data["sending_virus"] = sending_virus
+
+	return data
+
+////////////////////////
+// MESSAGE HANDLING
+////////////////////////
+
+// How I Learned To Stop Being A PDA Bloat Chump And Learn To Embrace The Lightweight
+
+// Gets the input for a message being sent.
+
+/datum/computer_file/program/messenger/proc/msg_input(mob/living/U = usr, rigged = FALSE)
+	var/t = tgui_input_text(U, "Enter a message", "NT Messaging")
+	if (!t || !sAndR)
+		return
+	if(computer.loc != U)
+		return
+	return t
+
+/datum/computer_file/program/messenger/proc/send_message(mob/living/user, list/obj/item/modular_computer/targets, everyone = FALSE, rigged = FALSE, fake_name = null, fake_job = null)
+	var/message = msg_input(user, rigged)
+	if(!message || !targets.len)
+		return FALSE
+	if((last_text && world.time < last_text + 10) || (everyone && last_text_everyone && world.time < last_text_everyone + 2 MINUTES))
```

Comment:
The `10` here should probably be a define, like, `MESSENGER_MESSAGE_DELAY`, and the 2 MINUTES should be `MESSENGER_SPAM_MESSAGE_DELAY`